calyx-opt/src/passes/static_promotion.rs

```diff
@@ -417,100 +476,270 @@ impl StaticPromotion {
+                    attributes,
+                    latency,
+                })
             }
-            _ => unreachable!("We do not insert non-static controls other than group enables with `promote_static` attribute")
-        }
+            ir::Control::Par(ir::Par { stmts, .. }) => {
+                // Convert stmts to static
+                let static_stmts =
+                    self.convert_vec_to_static(builder, std::mem::take(stmts));
+                // Calculate latency
+                let latency = static_stmts
+                    .iter()
+                    .map(|s| s.get_latency())
+                    .max()
+                    .unwrap_or_else(|| unreachable!("Empty Par Block"));
+                Self::check_latencies_match(latency, inferred_latency);
+                ir::StaticControl::Par(ir::StaticPar {
+                    stmts: static_stmts,
+                    attributes: ir::Attributes::default(),
+                    latency,
+                })
+            }
+            ir::Control::Repeat(ir::Repeat {
+                body, num_repeats, ..
+            }) => {
+                let sc = self.convert_to_static(body, builder);
+                let latency = (*num_repeats) * sc.get_latency();
+                Self::check_latencies_match(latency, inferred_latency);
+                ir::StaticControl::Repeat(ir::StaticRepeat {
+                    attributes: ir::Attributes::default(),
+                    body: Box::new(sc),
+                    num_repeats: *num_repeats,
+                    latency,
+                })
+            }
+            ir::Control::While(ir::While { body, .. }) => {
+                let sc = self.convert_to_static(body, builder);
+                let num_repeats = bound_attribute.unwrap_or_else(|| unreachable!("Called convert_to_static on a while loop without a bound"));
+                let latency = (num_repeats) * sc.get_latency();
+                Self::check_latencies_match(latency, inferred_latency);
+                ir::StaticControl::Repeat(ir::StaticRepeat {
+                    attributes: ir::Attributes::default(),
+                    body: Box::new(sc),
+                    num_repeats,
+                    latency,
+                })
+            }
+            ir::Control::If(ir::If {
+               port,  tbranch, fbranch, ..
+            }) => {
+                let static_tbranch = self.convert_to_static(tbranch, builder);
+                let static_fbranch = self.convert_to_static(fbranch, builder);
+                let latency = std::cmp::max(static_tbranch.get_latency(), static_fbranch.get_latency());
+                Self::check_latencies_match(latency, inferred_latency);
+                ir::StaticControl::static_if(Rc::clone(port), Box::new(static_tbranch), Box::new(static_fbranch), latency)
+            }
+            ir::Control::Static(_) => c.take_static_control(),
+            ir::Control::Invoke(_) => unreachable!(
```

Comment:
Why should invokes be compiled away before promotion?

Reply:
It was just bc `compile-invokes` was earlier in the pipeline than promotion. But I've added support for invokes now. 

On a side note, when/if we overhaul `with` and `comb groups` we should consider allowing `static invoke` to support comb groups; it make make sense in some situations.